File: LoadToCS.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"strconv"
	"strings"

	"github.com/360EntSecGroup-Skylar/excelize"
)

var rowNumCS = map[string]int{"Key": 0, "Type": 1, "DataStart": 4} // Key:字段key所在行 Type:字段类型所在行 DataStart:数据开始的行 (行数从0开始)

func LoadToCS(excelMap map[string]string) {
	entityHeaderCS := "using System.Collections;\nusing System.Collections.Generic;\n\n"
	entityResultCS := ""
	for name, path := range excelMap {
		dataResultMap := map[string]interface{}{}
		entityHeaderCS += "public class " + name + "   \n{\n"
		dataMap, entityStrCS := readExcel_CS(path)
		entityResultCS += entityStrCS
		for key, value := range dataMap {
			dataResultMap[key] = value
			entityHeaderCS += ("    public Dictionary<string," + key + "> " + key + ";\n")
		}
		byteBuf := bytes.NewBuffer([]byte{})
		encoder := json.NewEncoder(byteBuf)
		encoder.SetEscapeHTML(false)
		err := encoder.Encode(dataResultMap)
		if err == nil {
			saveData(byteBuf.String(), outPath+name+".json")
		} else {
			fmt.Println(err.Error())
		}
		entityHeaderCS += "}\n\n"
	}
	saveData(entityHeaderCS+entityResultCS, outPath+"DataEntity.cs")
}

func readExcel_CS(path string) (map[string]map[string]map[string]interface{}, string) {
	file, err := excelize.OpenFile(path)
	if err == nil {
		obj := map[string]map[string]map[string]interface{}{}
		entityStrCS := ""
		sheetMap := file.GetSheetMap()
		for _, sheetName := range sheetMap {
			keyArr := []string{}
			typeArrCS := []string{}
			sheetKey := strings.Title(sheetName)
			obj[sheetKey] = map[string]map[string]interface{}{}
			subObj := obj[sheetKey]
			rows := file.GetRows(sheetName)
			for i, row := range rows {
				for j, cellValue := range row {
					if i == rowNumCS["Key"] {
						keyArr = append(keyArr, cellValue)
					}
					if i == rowNumCS["Type"] {
						typeArrCS = append(typeArrCS, cellValue)
					}
					if i >= rowNumCS["DataStart"] {
						if j == 0 {
							subObj[cellValue] = map[string]interface{}{}
						}
						if typeArrCS[j] == "none" {
							continue
						}
						if typeArrCS[j] != "" {
							subObj[row[0]][keyArr[j]] = getValueByType_CS(cellValue, typeArrCS[j])
						}
					}
				}
			}
			entityStrCS += "public class " + sheetKey + "  \n{\n"
			if len(keyArr) == len(typeArrCS) {
				for k, v := range typeArrCS {
					if v == "none" || v == "" {
						continue
					}
					entityStrCS += "    public " + typeArrCS[k] + " " + keyArr[k] + ";\n"
				}
			}
			entityStrCS += "}\n\n"
		}
		return obj, entityStrCS
	} else {
		fmt.Println(err.Error())
	}
	return nil, ""
}

func getValueByType_CS(str string, typeStr string) interface{} {
	str = strings.Trim(str, ";")
	strArr := strings.Split(str, ";")
	var value interface{}
	var arr []interface{}
	if strings.Contains(typeStr, "boolean") {
		if strings.Contains(typeStr, "[]") {
			for _, v := range strArr {
				arr = append(arr, IF(v == "1", true, false))
			}
		} else {
			value = IF(value == "1", true, false)
		}
	} else if strings.Contains(typeStr, "number") {
		if strings.Contains(typeStr, "[]") {
			for _, v := range strArr {
				v1, _ := strconv.ParseFloat(v, 64)
				arr = append(arr, v1)
			}
		} else {
			v1, _ := strconv.ParseFloat(str, 64)
			value = v1
		}
	} else if typeStr == "string[]" {
		for _, v := range strArr {
			arr = append(arr, v)
		}
	} else {
		value = str
	}
	return IF(strings.Contains(typeStr, "[]"), arr, value)
}

File: LoadToCS.py
import LoadExcel
from xlrd.sheet import Sheet
import xlrd
import json

# Key:字段key所在行 Type:字段类型所在行 DataStart:数据开始的行 (行数从0开始)
rowNum = {"Key": 2, "Type": 1, "DataStart": 4}

def Load(excelDict: dict):
    entityHeader: str = "using System.Collections;\nusing System.Collections.Generic;\n\n"
    entityResult: str = ""
    for name in excelDict:
        dataResult = {}
        entityHeader += "public class " + name + "   \n{\n"
        wbResult = readExcel(excelDict[name])
        entityResult += wbResult["entityStr"]
        for key in wbResult["wbDict"]:
            dataResult[key] = wbResult["wbDict"][key]
            entityHeader += ("    public Dictionary<string," + key + "> " + key + ";\n")
        entityHeader += "}\n\n"
        LoadExcel.saveData(json.dumps(dataResult, ensure_ascii=False), name+".json")
    LoadExcel.saveData(entityHeader+entityResult, "DataEntity.cs")


def readExcel(path: str):
    with xlrd.open_workbook(path) as workbook:
        wbDict: dict = {}
        entityStr: str = ""
        sheetNames = workbook.sheet_names()
        for sheetName in sheetNames:
            print("load sheet "+sheetName+" start")
            sName = LoadExcel.upperFirst(sheetName)
            wbDict[sName] = {}
            sheetDict = wbDict[sName]
            keyList: list = []
            typeList: list = []
            sheet: Sheet = workbook.sheet_by_name(sheetName)
            for row in range(sheet.nrows):
                idStr = ""
                for col in range(sheet.ncols):
                    cellV = str(sheet.cell_value(row, col)).rstrip("0")
                    cellV = cellV.rstrip(".")
                    if row == rowNum["Key"]:
                        keyList.insert(col, cellV)
                    if row == rowNum["Type"]:
                        typeList.insert(col, cellV)
                    if row >= rowNum["DataStart"]:
                        if col == 0:
                            sheetDict[str(cellV)] = {}
                            idStr = cellV
                        if typeList[col] == "none" or typeList[col] == "":
                            continue
                        sheetDict[idStr][keyList[col]] = getValueByType(
                            cellV, typeList[col])
            entityStr += "public class " + sName + "  \n{\n"
            if len(keyList) == len(typeList):
                for i in range(len(typeList)):
                    v = typeList[i]
                    if v == "none" or v == "":
                        continue
                    entityStr += "    public " + typeList[i] + " " + keyList[i] + ";\n"
                entityStr += "}\n\n"
            print("load sheet "+sheetName+" end")
        return {"entityStr": entityStr, "wbDict": wbDict}


def getValueByType(cellV: str, typeStr: str):
    cellV = cellV.strip()
    cellV = cellV.strip(";")
    strList = cellV.split(";")
    value = None
    arr: list = []
    if typeStr.find("bool") > -1:
        if typeStr.find("[]") > -1:
            for v in strList:
                arr.append(LoadExcel.IF(v == 1, True, False))
        else:
            value = LoadExcel.IF(cellV == 1, True, False)
    elif typeStr.find("int") > -1 or typeStr.find("float") > -1:
        if typeStr.find("[]") > -1:
            for v in strList:
                try:
                    arr.append(LoadExcel.IF(typeStr.find("float") > -1, float(v), int(v)))
                except Exception:
                    arr.append(0)
        else:
            try:
                value = LoadExcel.IF(typeStr.find("float") > -1, float(cellV), int(cellV))
            except Exception:
                value = 0

    elif typeStr.find("string") > -1:
        if typeStr.find("[]") > -1:
            for v in strList:
                arr.append(v)
        else:
            value = cellV
    else:
        print("不支持的数据类型", typeStr)
    return LoadExcel.IF(typeStr.find("[]") > -1, arr, value)
